python/lsst/daf/butler/core/utils.py

```diff
@@ -290,3 +292,93 @@ def __iter__(self):
         if _nConnections != 0:
             raise ValueError('Cycle found')
         return
+
+
+class TopologicalSet:
+    """A collection that behaves like a builtin `set`, but where
+    elements can be interconnected (like a graph).
+
+    Iteration over this collection visits its elements in topologically
+    sorted order.
+
+    Parameters
+    ----------
+    elements : `iterable`
+        Any iterable with elements to insert.
+    """
+    Node = namedtuple('Node', ['element', 'sourceElements'])
+
+    def __init__(self, elements):
+        self._nodes = {e: TopologicalSet.Node(e, set()) for e in elements}
+        self._ordering = None
+
+    def __contains__(self, element):
+        return element in self._nodes
+
+    def __len__(self):
+        return len(self._nodes)
+
+    def connect(self, sourceElement, targetElement):
+        """Connect two elements in the set.
+
+        The connection is directed from `sourceElement` to `targetElement` and
+        is distinct from its inverse.
+        Both elements must already be present in the set.
+
+        sourceElement : `object`
+            The source element.
+        targetElement : `object`
+            The target element.
+
+        Raises
+        ------
+        KeyError
+            When either element is not already in the set.
+        ValueError
+            If a self connections between elements would be created.
+        """
+        if sourceElement == targetElement:
+            raise ValueError('Cannot connect {} to itself'.format(sourceElement))
+        for element in (sourceElement, targetElement):
```

Comment:
I still think:
`
missing = {sourceElement, TargetElement} - self._elements 
if missing: raise KeyError('element(s) {x} not in set'.format(x=missing))
`
Would be better than a for loop. It would also have the benefit of telling the user if point points are invalid rather than bailing at the first